lgc/util/llpcGfxRegHelper.h

```diff
@@ -0,0 +1,153 @@
+/*
+ ***********************************************************************************************************************
+ *
+ *  Copyright (c) 2020 Advanced Micro Devices, Inc. All Rights Reserved.
+ *
+ *  Permission is hereby granted, free of charge, to any person obtaining a copy
+ *  of this software and associated documentation files (the "Software"), to deal
+ *  in the Software without restriction, including without limitation the rights
+ *  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ *  copies of the Software, and to permit persons to whom the Software is
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ *
+ **********************************************************************************************************************/
+ /**
+  ***********************************************************************************************************************
+  * @file  llpcGfxRegHelper.h
+  * @brief LLPC header file: contains the definition of LLPC utility class GfxRegHelper.
+  *
+  * @details The class handles symbolic expressions over values read from hardware graphics registers, materialized as
+  *          llvm instructions. The code maintains a map from a register ID to an llvm value with the current symbolic
+  *          expression describing the updated state of the register.
+  *
+  * @sa llpc/docs/DdnGraphicsRegisterHandler.md
+  ***********************************************************************************************************************
+  */
+
+#pragma once
+#include "llpcGfxRegHandler.h"
+#include "llpcTargetInfo.h"
+
+namespace lgc
+{
+
+using namespace llvm;
+
+// =====================================================================================================================
+// The goal of the helper structs is to maintain the newest symbolic state of a register and eventually produce
+// optimized llvm instruction sequences for these states.
+// - pValue : Cache the value which is already being fetched.
+// - isModified : True, if the register is being settled with new value.
```

Comment:
In general, I don't think you can say that 'a register is settled with a value'. I may be wrong here, but it's not a standard term from what I know, and it'd be better to define it directly.
Maybe say instead that: 'This indicates that the value read from a hardware register was given a new symbolic state (on top of a freshly read register value). Note that we don't actually modify the state of hardware registers.

Reply:
I just droped the previous comments and added a state table on it.